### PR TITLE
[NRH-30] Stripe Onboarding

### PR DESCRIPTION
**Pull request by michael-caktus**
The primary work for this ticket was implementing Stripe onboarding.
This effort required two new views:

1. `stripe_onboarding` - creates a stripe.Account and a stripe.AccountLink, which provides a URL to direct the user to the onboarding flow.
2. `stripe_confirmation` - Because the onboarding process can be interrupted at any time, or simply not completed, another request to Stripe is required to verify validity of the account created. This same request can be used to verify/re-verify a stripe account. A "verified" stripe account can accept payments.

The front end flow to manage all of this is pretty complicated, but in short:
1. HubAdmin creates an Organization and attaches a new User to that org.
    - for now, they make up a password for that user. Eventually, OrgAdmin will need to go through the "forgot-password flow" to set their password
2. OrgAdmin logs in to dashboard.
3. OrgAdmins see a view requiring them to connect a payment provider before continuing to the rest of the dashboard.
4. OrgAdmin can click on a Connect with Stripe button that kicks off the onboarding flow.
5. Stripe onboarding continues in another tab
6. When OrgAdmin returns, request is made to `stripe_confirmation` to check status of stripe account connection.

Lots of extra stuff has been added in this PR as well. The front-end needed some actual design choices made, and various ancillaries including error/success messaging, routing, etc.

Also brought in cypress for front end testing.  This PR attempts to integrate cypress in to Github CI. Expect iteration...

### CODE REVIEW

File: spa/src/components/content/Content.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        <p>Content 1</p>
+      </S.ContentSection>
+      <S.ContentSection>
+        <p>Content 2</p>
+      </S.ContentSection>
+      <S.ContentSection>
+        <p>Content 3</p>
+      </S.ContentSection>
+    </S.Content>
+  );
```
Comment (benjaminEwhite) [inline spa/src/components/content/Content.js:16]: I gather this is just placeholder for now, no?

File: spa/src/components/dashboard/Dashboard.styled.js [hunk 2]
```diff
@@ -4,6 +4,21 @@ export const Dashboard = styled.div`
...
-  align-items: center;
-  justify-content: space-around;
+`;
+
+export const DashBody = styled.div`
+  flex: 1;
+  display: flex;
+  flex-direction: row;
+
+  @media (${(p) => p.theme.breakpoints.tabletLandscapeDown}) {
```
Comment (benjaminEwhite) [inline spa/src/components/dashboard/Dashboard.styled.js:14]: ditto

File: spa/src/components/dashboard/DashboardSection.styled.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import styled from 'styled-components';
+
+export const DashboardSection = styled.section`
+  border-radius: ${(p) => p.theme.radii[1]};
+  box-shadow: ${(p) => p.theme.shadows[1]};
+  background: ${(p) => p.theme.colors.paneBackground};
```
Comment (benjaminEwhite) [inline spa/src/components/dashboard/DashboardSection.styled.js:6]: Ditto

File: spa/src/components/dashboard/DashboardSectionGroup.styled.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import styled from 'styled-components';
+import { motion } from 'framer-motion';
+
+export const DashboardSectionGroup = styled(motion.ul)`
+  padding: 0;
+  margin: 0 auto;
+  width: 100%;
+  max-width: ${(p) => p.theme.maxWidths.lg};
```
Comment (benjaminEwhite) [inline spa/src/components/dashboard/DashboardSectionGroup.styled.js:8]: Ditto. Going to stop saying ditto from hereon out though.

**Review by benjaminEwhite — COMMENTED** [review]
Maybe the size of this diff is what broke syntax highlighting in the "Files Changed" view of this PR. At any rate, this is *huge*, and it's probably worth saying that I lost a bit of steam part way through the SPA code (right around where syntax highlighting went away).


Left a bunch of inline comments, mainly on server-side code. 

A few high level comments:

- I had some trouble grokking what was going on in apps/contributions/tests/test_views.py without spending time doing lots of cross-referencing between object code and test code. It might be a good idea to include doc strings to describe what's being tested/why there, to help out future maintainers.
- Ideally, relevant server-side tests would use the same fixtures that Cypress is using, as this would be a better guarantee of a contract between front and backend.
- There is a TON of SPA code in this PR, with a ton of logic in it, and it would seem that very little of it is tested. I'm somewhat of a skeptic about extensive front end testing for a variety of reasons, and am not saying we should test all the branches through the front end code, but I will say that it's pretty overwhelming to read through all the front end code and try to grok what's happening. That is arguably inevitable for a React app with a minimum of complexity, but I guess I am slightly concerned about the maintainability of this for future maintainers who are not Caktus.

To be clear, not trying to be discouraging -- this overall looks good, it just happens to be a very big, and complex diff.

File: apps/contributions/views.py [hunk 5]
```diff
@@ -21,9 +21,9 @@ def convert_money_value_to_stripe_payment_amount(amount):
     return int(float(amount) * 100)
 
 
-@api_view(["POST", "PATCH"])
+@api_view(["POST"])
 def stripe_payment_intent(request):
-    if request.method == "POST":
+    if request.method == "POST":  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (michael-caktus) [reply apps/contributions/views.py:26]: I'd meant to ignore coverage on the implied "else" there, but it looks like I ignored that entire branch of logic instead. That'll be fixed.

File: apps/contributions/views.py [hunk 6]
```diff
@@ -70,6 +70,81 @@ def stripe_payment_intent(request):
...
+@api_view(["POST"])
+def stripe_onboarding(request):
+    if request.method == "POST":  # pragma: no cover
+        organization = request.user.get_organization()
+        company = pre_populate_account_company_from_org(organization)
+
+        try:
+            account = stripe.Account.create(
+                type="standard",
+                company=company,
```
[the remark this answers is not in the document]
Comment (michael-caktus) [reply apps/contributions/views.py:95]: Fixed the error mentioned above-- but, yes, Stripe does, but here "company" for Stripe just represents data about the company like "name" and stuff.

File: apps/users/models.py [hunk 7]
```diff
@@ -31,6 +31,9 @@ def get_full_name(self):
     def get_short_name(self):
         return self.email
 
+    def get_organization(self):
+        return self.organization_set.first()
```
[the remark this answers is not in the document]
Comment (michael-caktus) [reply apps/users/models.py:35]: We don't actually support members belonging to multiple orgs--though we've coded it to support it in the future. I've added a commend to this method explaining.

File: spa/src/ajax/axios.js [hunk 8]
```diff
@@ -34,16 +34,11 @@ Axios.interceptors.request.use(
 Axios.interceptors.response.use((success) => success, handleResponseError);
 
 function handleResponseError(error) {
-  if (error?.response) {
-    const { status } = error.response;
-    // Only care about 403s so far, so pass through
-    if (status !== 403) return Promise.reject(error);
-    return handle403Response(error);
-  }
+  if (error?.response?.status === 401) return handle401Response(error);
```
[the remark this answers is not in the document]
Comment (michael-caktus) [reply spa/src/ajax/axios.js:37]: Well, specifically-- if the server returns a "We don't have credentials for you anymore" they get de-authed. Different from 403 Forbidden, which we'll get if they don't have permission. 401 should only be returned when they no longer have a valid token. This gets hit when their JWT expires, forcing them to log back in.

File: apps/contributions/views.py [hunk 9]
```diff
@@ -21,53 +21,125 @@ def convert_money_value_to_stripe_payment_amount(amount):
...
+        organization.save()
+
+        account_links = stripe.AccountLink.create(
+            account=account.id,
+            refresh_url=f"{settings.SITE_URL}?cb=stripe_reauth",
+            return_url=f"{settings.SITE_URL}?cb=stripe_return",
+            type="account_onboarding",
+            api_key=get_hub_stripe_api_key(),
+        )
+    except stripe.error.StripeError:
```
Comment (benjaminEwhite) [inline apps/contributions/views.py:107]: Maybe not worthing considering, but is there an edge case where part of this block succeeds but not the whole thing, and that it gets retriggered a second time?  Just wondering what would happen if account gets created more than once. At a minimum, are we getting logs the Stripe library by calling its methods?

File: apps/contributions/views.py [hunk 10]
```diff
@@ -21,53 +21,125 @@ def convert_money_value_to_stripe_payment_amount(amount):
...
+            return Response({"status": "not_connected"}, status=status.HTTP_202_ACCEPTED)
+        # A previously confirmed account can spare the stripe API call
+        if organization.stripe_verified:
+            return Response({"status": "connected"}, status=status.HTTP_200_OK)
+
+        # A "Confirmed" stripe account has "charges_enabled": true on return from stripe.Account.retrieve
+        stripe_account = stripe.Account.retrieve(organization.stripe_account_id, api_key=get_hub_stripe_api_key())
+
+    except stripe.error.StripeError:
+        # ? Send email?
```
Comment (benjaminEwhite) [inline apps/contributions/views.py:131]: Seems prudent

Comment (benjaminEwhite) [reply apps/contributions/views.py:131]: Probably should at least log for now.

File: spa/src/components/connect/stripe/StripeConnect.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+      setIsLoading(false);
+      let alertMessage = "Something went wrong! We've been notified";
+      if (e.response?.data?.detail) alertMessage = e.response.data.detail;
+      alert.error(alertMessage, { timeout: 8000 });
+    }
+  }, [isLoading, providerConnectInProgress, alert, setProviderConnectInProgress]);
+
+  return (
+    <S.StripeConnect onClick={handleStripeOnboarding} isLoading={isLoading} data-testid="stripe-connect-button">
+      {isLoading ? <Spinner /> : <span>Connect with</span>}
```
Comment (benjaminEwhite) [inline spa/src/components/connect/stripe/StripeConnect.js:40]: I guess there's a stripe icon adjacent to "Connect with"?  Would be good to have text here as well, even if it's not visibly displayed.

**Review by benjaminEwhite — APPROVED** [review]
A couple of small comments, but no deal breakers. LGTM.

**Pull request merged**